src/adapter/src/coord/sequencer/inner.rs

```diff
@@ -4631,6 +4608,295 @@ impl Coordinator {
+        }
+    }
+
+    async fn sequence_alter_connection_options(
+        &mut self,
+        mut ctx: ExecuteContext,
+        id: GlobalId,
+        set_options: BTreeMap<ConnectionOptionName, Option<WithOptionValue<mz_sql::names::Aug>>>,
+        drop_options: BTreeSet<ConnectionOptionName>,
+        validate: bool,
+    ) {
+        let cur_entry = self.catalog().get_entry(&id);
+        let cur_conn = cur_entry.connection().expect("known to be connection");
+
+        let inner = || -> Result<catalog::Connection, AdapterError> {
+            // Parse statement.
+            let create_conn_stmt = match mz_sql::parse::parse(&cur_conn.create_sql)
+                .expect("invalid create sql persisted to catalog")
+                .into_element()
+                .ast
+            {
+                Statement::CreateConnection(stmt) => stmt,
+                _ => unreachable!("proved type is source"),
+            };
+
+            let catalog = self.catalog().for_system_session();
+
+            // Resolve items in statement
+            let (mut create_conn_stmt, resolved_ids) =
+                mz_sql::names::resolve(&catalog, create_conn_stmt)
+                    .map_err(|e| AdapterError::internal("ALTER CONNECTION", e))?;
+
+            // Retain options that are neither set nor dropped.
+            create_conn_stmt
+                .values
+                .retain(|o| !set_options.contains_key(&o.name) && !drop_options.contains(&o.name));
+
+            // Set new values
+            create_conn_stmt.values.extend(
+                set_options
+                    .into_iter()
+                    .map(|(name, value)| ConnectionOption { name, value }),
+            );
+
+            // Open a new catalog, which we will use to re-plan our
+            // statement with the desired config.
+            let mut catalog = self.catalog().for_system_session();
+            catalog.mark_id_unresolvable_for_replanning(id);
+
+            // Re-define our source in terms of the amended statement
+            let plan = match mz_sql::plan::plan(
+                None,
+                &catalog,
+                Statement::CreateConnection(create_conn_stmt),
+                &Params::empty(),
+                &resolved_ids,
+            )
+            .map_err(|e| AdapterError::InvalidAlter("CONNECTION", e))?
+            {
+                Plan::CreateConnection(plan) => plan,
+                _ => unreachable!("create source plan is only valid response"),
+            };
+
+            // Parse statement.
+            let create_conn_stmt = match mz_sql::parse::parse(&plan.connection.create_sql)
+                .expect("invalid create sql persisted to catalog")
+                .into_element()
+                .ast
+            {
+                Statement::CreateConnection(stmt) => stmt,
+                _ => unreachable!("proved type is source"),
+            };
+
+            let catalog = self.catalog().for_system_session();
```

Comment:
`mz_sql::names::resolve` takes a `SessionCatalog`/`ConnCatalog` as its argument, so we have to provide one of these--because we'd already mucked with the previous catalog we generated, I figured just getting a pristine new catalog was less likely to cause issues.